backends/p4tools/testgen/targets/bmv2/table_stepper.cpp

```diff
@@ -0,0 +1,326 @@
+#include "backends/p4tools/testgen/targets/bmv2/table_stepper.h"
+
+#include <cstddef>
+#include <map>
+#include <ostream>
+#include <string>
+#include <vector>
+
+#include <boost/none.hpp>
+#include <boost/optional/optional.hpp>
+
+#include "backends/p4tools/common/lib/formulae.h"
+#include "backends/p4tools/common/lib/ir.h"
+#include "backends/p4tools/common/lib/trace_events.h"
+#include "lib/error.h"
+#include "lib/null.h"
+#include "lib/safe_vector.h"
+
+#include "backends/p4tools/testgen/lib/continuation.h"
+#include "backends/p4tools/testgen/lib/exceptions.h"
+#include "backends/p4tools/testgen/options.h"
+#include "backends/p4tools/testgen/targets/bmv2/constants.h"
+
+namespace P4Tools {
+
+namespace P4Testgen {
+
+namespace Bmv2 {
+
+const IR::Expression* BMv2_V1ModelTableStepper::computeTargetMatchType(
+    ExecutionState* nextState, const KeyProperties& keyProperties,
+    std::map<cstring, const FieldMatch>* matches, const IR::Expression* hitCondition) {
+    const IR::Expression* keyExpr = keyProperties.key->expression;
+
+    // For now, we consider optional match types to be a no-op.
+    if (keyProperties.matchType == BMv2Constants::MATCH_KIND_OPT) {
+        return hitCondition;
+    }
+    // Action selector entries are not part of the match.
+    if (keyProperties.matchType == BMv2Constants::MATCH_KIND_SELECTOR) {
+        bmv2_V1ModelProperties.actionSelectorKeys.emplace_back(keyExpr);
+        return hitCondition;
+    }
+    // Ranges are not yet implemented for Tofino STF tests.
```

Comment:
The reason why we check the test framework here is so that we can fail early instead of stepping through the entire program. Also in some cases, we can have a failover mechanism if the test framework does not support a specific construct.